org.lflang/src/org/lflang/graph/InstantiationGraph.java

```diff
@@ -0,0 +1,178 @@
+        } else {
+            return CollectionLiterals.<Instantiation>emptySet();
+        }
+    }
+
+    /**
+     * Return the declarations that point to a given reactor definition.
+     * A declaration is either a reactor definition or an import statement.
+     */
+    public Set<ReactorDecl> getDeclarations(final Reactor definition) {
+        return this.reactorToDecl.get(definition);
+    }
+
+    /**
+     * Return the reactor definitions referenced by instantiations in this graph
+     * ordered topologically. Each reactor in the returned list is preceded by
+     * any reactors that it may instantiate.
+     */
+    public List<Reactor> getReactors() {
+        return this.nodesInTopologicalOrder();
+    }
+
+    /**
+     * Construct an instantiation graph based on the given AST and, if the
+     * detectCycles argument is true, run Tarjan's algorithm to detect cyclic
+     * dependencies between instantiations.
+     * @param resource The resource associated with the AST.
+     * @param detectCycles Whether or not to detect cycles.
+     */
+    public InstantiationGraph(final Resource resource, final boolean detectCycles) {
+        final Iterable<Instantiation> instantiations = Iterables.<Instantiation>filter(
+            IteratorExtensions.<EObject>toIterable(resource.getAllContents()), Instantiation.class);
+        final Reactor main = IterableExtensions.<Reactor>findFirst(
+            Iterables.<Reactor>filter(
+                IteratorExtensions.<EObject>toIterable(resource.getAllContents()), Reactor.class),
+            (Reactor it) -> {
+                return Boolean.valueOf((it.isMain() || it.isFederated()));
+            });
+        if (main != null) {
+            this.addNode(main);
+        }
+        for (final Instantiation i : instantiations) {
+            this.buildGraph(i, CollectionLiterals.<Instantiation>newHashSet());
+        }
+        if (detectCycles) {
+            this.detectCycles();
+        }
+    }
+
+    /**
+     * Construct an instantiation graph based on the given AST and, if the
+     * detectCycles argument is true, run Tarjan's algorithm to detect cyclic
+     * dependencies between instantiations.
+     * @param model The root of the AST.
+     * @param detectCycles Whether or not to detect cycles.
+     */
+    public InstantiationGraph(final Model model, final boolean detectCycles) {
+        for (final Reactor r : model.getReactors()) {
+            for (final Instantiation i : r.getInstantiations()) {
+                this.buildGraph(i, CollectionLiterals.<Instantiation>newHashSet());
+            }
+        }
+        if (detectCycles) {
+            this.detectCycles();
+        }
+    }
+
+    /**
+     * Traverse the AST and build this precedence graph relating the
+     * encountered instantiations. Also map each reactor to all
+     * declarations associated with it and each reactor to the sites of
+     * its instantiations.
+     * @param instantiation
+     * @param graph
```

Comment:
The `graph` parameter is gone now. This is private, so just deleting the `@param` comments might be an acceptable solution.
```suggestion
```